### PR TITLE
add "lighthouse2" effect

### PR DESCRIPTION
**Pull request by Mar-Pfa**
creates moving dot that can move e.g. around a circle also with support of multiple rows and smoooth transition
lighthouse (1)

### CODE REVIEW

File: wled00/FX.cpp
Comment (blazoncek) [inline wled00/FX.cpp:406]: Why 36? Not everyone wants a row length of 36.

File: wled00/FX.cpp
Comment (blazoncek) [inline wled00/FX.cpp:413]: Use float. It is faster and has enought resolution.

File: wled00/FX.cpp
Comment (blazoncek) [inline wled00/FX.cpp:414]: Use WLED math functions. They are accurate & fast enough but use less memory.

File: wled00/FX.cpp
Comment (blazoncek) [inline wled00/FX.cpp:440]: Use color_blend().

File: wled00/FX.cpp
Comment (blazoncek) [inline wled00/FX.cpp:459]: Recommended is to use SEGCOLOR(1) instead.

**Review by blazoncek — CHANGES_REQUESTED** [review]
IMO this FX reqiures special set-up to be really effective.
Please consider rewriting it using comments and think about how to make it length independent.

**Comment by Mar-Pfa** [comment]
Thanks for the Feedback! I improved and simplified the code a lot. Only thing i was not able to bring in was the wled_math.h. I wanted to use sin_t / cos_t - when adding wled_math.h to to fx.cpp i get build errors because of dublication. without adding the library i get build errors as the functions are not know :-(
Any ideas on that?

File: wled00/FX.cpp
Comment (Mar-Pfa) [reply wled00/FX.cpp]: changed it to be configurable with parameter 2, now the dotwidth is fixed for now (but will be variable again with next version)

Comment (blazoncek) [reply wled00/FX.cpp]: Taking 3rd parameter will be available in 2D version of WLED sometime after summer.

**Comment by blazoncek** [comment]
You may need to add
```
#include "wled.h"
#include "fcn_declare.h"
```
to FX.cpp

**Comment by Mar-Pfa** [comment]
:-( doesn't work...  wled.h is already in, adding fcn-declare.h does not help. the build problem is:

.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld.exe: .pio\build\d1_mini_ota\src\ntp.cpp.o: in function `cos_t(float)':
ntp.cpp:(.text._Z5cos_tf+0xb4): multiple definition of `cos_t(float)'; .pio\build\d1_mini_ota\src\FX.cpp.o:FX.cpp:(.text._Z5cos_tf+0xb4): first defined here
.platformio/packages/toolchain-xtensa/bin/../lib/gcc/xtensa-lx106-elf/4.8.2/../../../../xtensa-lx106-elf/bin/ld.exe: .pio\build\d1_mini_ota\src\ntp.cpp.o: in function `sin_t(float)':
...

don't know how to fix this

At least -> with using just the plain "sin(x)" method it works so maybe we can still merge my pull-request!? getting this wled_math.h to work in FX.cpp might be more effort!?

**Comment by blazoncek** [comment]
> don't know how to fix this

Oh, yes, it requires wled_math be a .cpp instead of .h. A known issue in current main branch.
You may need to wait until #2610 gets merged.
sin() and cos() will not be merged.

**Comment by Mar-Pfa** [comment]
@[blazoncek](https://github.com/blazoncek) finally found an alternative :-) I was not aware of the sin16 / cos16 - so fixed the code, now it works and without sin() / cos() - hopefully it's fine now!?

**Review by blazoncek — APPROVED** [review]

**Comment by blazoncek** [comment]
You may want to test-drive 2D implementation from my fork which will unfortunately make this PR obsolete if you set up your lights as 2D matrix.
Otherwise this is ok.

**Comment by blazoncek** [comment]
I can. 😄 But will wait for @Aircoookie to peer review and approve (may take a while though).
Adding new effects is a delicate matter since it is very annoying (for users) if it gets removed later on due to duplicates.

**Comment by Aircoookie** [comment]
Thank you for the great PR (and lamp!) and sorry for taking so long to review it.
You couldn't realistically have anticipated 2D support when you made this PR, so I apologize.
In 0.14, I personally think configuring the light as a matrix and using e.g. the standard lighthouse effect with "Bar" 1D->2D mapping would suffice.
Still I really like the smooth fade in and out, I think it could serve as a nice base for a combined 1D and 2D effect - nice smooth fade on 1D and a circular "fresnel lens" type of thing moving across the matrix in 2D, so I will leave this open for now :)

**Pull request closed without merging**
